src/Expr.cc

```diff
@@ -1815,7 +1815,7 @@ ValPtr EqExpr::Fold(Val* v1, Val* v2) const
 	{
 	if ( op1->GetType()->Tag() == TYPE_PATTERN )
 		{
-		RE_Matcher* re = v1->AsPattern();
+		auto re = dynamic_cast<PatternVal*>(v1);
```

Comment:
We have a few of such unchecked `dynamic_cast` now. I suggest we add a checked, templated `as<T>()` method to `Val` instead.

Reply:
We actually already have a bunch of type-checking `As` methods, such as `AsPatternVal`.. Is there a reason why these aren't just being used here?

Reply:
I added the templated `As()` method in https://github.com/zeek/zeek/commit/85164cd67b66706306c293f1ff1033f36f49b838, but the question about the existing `As` method above still stands.

Reply:
The reason I didn't do this with `AsPatternVal` is that the original code was written in terms of `re` being an `RE_Matcher`, not a `PatternVal`.  Because `PatternVal` has a pass-through `MatchExactly` method, however, your proposed change of `As<PatternVal*>` works fine.

Reply:
Sounds good to me. I'll mark this as resolved then.